src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs

```diff
@@ -29,17 +29,32 @@ internal static class ThreadPoolGlobals
 
         public static readonly ThreadPoolWorkQueue workQueue = new ThreadPoolWorkQueue();
 
-        /// <summary>Shim used to invoke <see cref="IAsyncStateMachineBox.MoveNext"/> of the supplied <see cref="IAsyncStateMachineBox"/>.</summary>
-        internal static readonly Action<object?> s_invokeAsyncStateMachineBox = state =>
+        /// <summary>Shim used to invoke <see cref="StateMachineWorkItem.MoveNext"/> of the supplied <see cref="StateMachineWorkItem"/>.</summary>
+        internal static readonly Action<object?> s_invokeStateMachineWorkItem = Invoke.Instance.StateMachineWorkItem;
+
+        private class Invoke
         {
-            if (!(state is IAsyncStateMachineBox box))
+            // Invoking an delegate on an instance is faster than invoking it as a static method.
+            public static Invoke Instance { get; } = new Invoke();
+
+            public void StateMachineWorkItem(object? state)
             {
-                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.state);
-                return;
+                if (!(state is StateMachineWorkItem box))
+                {
+                    ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.state);
+                    return;
+                }
+
+                box.MoveNext();
```

Comment:
Seems like this draft is:
- Reducing a cast cost by casting to a class instead of an interface
- Increasing cost by adding another virtual dispatch
How do those balance out?

And then separately there's the fairly unrelated change to how the delegate is created/cached.  If that has a measurable impact, it should be separated out so as not to be muddled in with the primary thing being experimented with here.

Reply:
Ah, actually, I misread the change in IThreadPoolWorkItem.MoveNext.  I guess it's not actually adding a virtual dispatch there on the fast path.

Reply:
Yeah, looking to move that out as a separate change as it works on its own merits https://github.com/dotnet/runtime/pull/36538

Not convinced by the rest of the changes; isn't a fast check to typeof with an open generic which I thought there was initially